### ISSUE TITLE
Is there a way to suppress text interpolation for strings containing regexes?

### ISSUE THREAD

**Issue by coderfromhere** [dhall-lang/dhall-lang#1290]
Hi, I've got a text value of a regular expression that I want to pass verbatim to downstream formats. Let's say it's something like `"^[0-9]{4}\.$"` How can I suppress string interpolation for text values in that case?

**Comment by Gabriella439** [comment]
There is not a way to disable string interpolation for text literals, but you can escape it, using either `\${` (for double-quoted `Text` literals) or `''${` (for multi-line `Text` literals)

**Comment by coderfromhere** [comment]
I've also found that if I make it multi-line, in some formats like YAML it does produce a literal value: of a form `val: |\n<literal>`, but I wonder if it's consistent across all possible formats (how does it behave with JSON?)

**Comment by Gabriella439** [comment]
It's guaranteed to be consistent across all formats, because multi-line `Text` literals are syntactic sugar for double-quoted `Text` literals, meaning that multi-line `Text` literals are converted to double-quoted `Text` literals before any downstream processing occurs (e.g. conversion to JSON/YAML)

**Comment by coderfromhere** [comment]
ok, if it's just syntatic sugar for double-quoted literal, I'm a bit puzzled why the example above works for me, even though I didn't put explicit `''` right next to the line containing the regex. Is it looking backward to find an opening `''` while parsing the interpolation?

**Comment by Gabriella439** [comment]
If I remember correctly, the `dhall-to-yaml` and `dhall-to-yaml-ng` tools only look for the presence of a newline in the `Text` value to decide whether or not to render as a multi-line or single-line YAML value.

So what happen is that:

* The Dhall multi-line `Text` literal is desugared to a double-quoted `Text` literal containing a newline (e.g. `"regex:^[0-9]{4}\.$\n"`)
* The `dhall-to-yaml` tool see the `\n` in the `Text` literal and knows to render as a multi-line YAML value

You can verify this yourself by passing in a double-quoted `Text` literal with an explicit escaped newline into the `dhall-to-yaml` tool and it should behave the same way

**Comment by coderfromhere** [comment]
Interesting, I'm getting a different behaviour based on a string literal constructor:

1) Double-quoted (with/without a leading newline sequence):
```
let serviceTagFlux  = \(svcName: Text) ->
    {   mapKey      = "fluxcd.io/tag.${svcName}"
    ,   mapValue    = "\nregexp:^[0-9]{4}\.$"
    }
```
produces parsing error
```
    |
166 |     ,   mapValue    = "\nregexp:^[0-9]{4}\.$"
    |         ^
unexpected 'm'
expecting whitespace or }
```

2) multi-line literal
```
let serviceTagFlux  = \(svcName: Text) ->
    {   mapKey      = "fluxcd.io/tag.${svcName}"
    ,   mapValue    = ''
                      regexp:^[0-9]{4}\.$
                      ''
    }
```
produces YAML result
```
fluxcd.io/tag.mydeployment: |
      regexp:^[0-9]{4}\.$
```

**Comment by Gabriella439** [comment]
Yeah, the parsing error message is not great because of a Haskell-implementation-specific reason (see: https://github.com/dhall-lang/dhall-haskell/issues/2264), but the root of the problem is this:

```
$ dhall <<< '"\nregexp:^[0-9]{4}\.$"'
dhall: 
Error: Invalid input

(input):1:21:
  |
1 | "\nregexp:^[0-9]{4}\.$"
  |                     ^
unexpected '.'
expecting ", $, /, \, b, f, n, r, t, or u
```

According to [the standard grammar](https://github.com/dhall-lang/dhall-lang/blob/4bedfcfd409ca9914b435ab366e47b64b9a9ddf4/standard/dhall.abnf#L237-L261) you have to escape the `\`, so what you want is:

```dhall
let serviceTagFlux  = \(svcName: Text) ->
    {   mapKey      = "fluxcd.io/tag.${svcName}"
    ,   mapValue    = "\nregexp:^[0-9]{4}\\.$"
    }
```

**Comment by coderfromhere** [comment]
There's something off. If I use your last snippet:
```
let serviceTagFlux  = \(svcName: Text) ->
    {   mapKey      = "fluxcd.io/tag.${svcName}"
    ,   mapValue    = "\nregexp:^[0-9]{4}\\.$"
    }
```

and use it together with `dhall-kubernetes` `v1.21`, the generated yaml contains:
```
fluxcd.io/tag.mydeployment: "\nregexp:^[0-9]{4}\\.$"
```

i.e. it doesn't escape the value and prints it verbatim with all escape characters, whereas my goal is to get the original unescaped regex like in the example output of the multi-line string syntax above.

**Comment by coderfromhere** [comment]
Thanks for the hint, I'm going to switch to the ng one for my current goal!

But I'd still like to understand (and make it available for search engines' indexes, thanks to this discussion) why the behaviour of `dhall-to-yaml` specifically differes based on multi-line vs single-line constructors? What I mean is, let's consider a pipeline where:

* `dhall-to-yaml` is an invariant property of the pipeline
* the pipeline feeds the result of a compilation to a Python script that uses the regex value from YAML to construct a regular expression template:
  ```python
  import yaml
  x = next(yaml.safe_load_all(open('./.local/tmp.yaml', 'r'))).strip()
  # x == 'regexp:^[0-9]{4}\\.$'
  import re
template =  re.compile(x)
# use the template to perform other tasks down the pipeline stream
  ```

In this setting, the behaviour of the pipeline changes significantly due to a semantically insignificant string literal syntax swap somewhere up the pipeline stream.

Is this an intentional behaviour of the tool that is useful in other use cases or a property of the existing implementation of `dhall-to-yaml` that may not necessarily be correct and intended?

**Comment by Gabriella439** [comment]
My understanding is that the behavior of downstream YAML-dependent programs (e.g. the Python script you gave) would not change as a result of semantically insignificant changes.  The reason why is that:

```yaml
mapKey: fluxcd.io/tag.foo
mapValue: |

  regexp:^[0-9]{4}\.$
```

… and:

```yaml
mapKey: fluxcd.io/tag.foo
mapValue: "\nregexp:^[0-9]{4}\\.$\n"
```

… are semantically the same YAML configuration.

The YAML spec is actually unclear on this point, but given that it says that different flow scalar styles are not supposed to convey content information the inference I've drawn from that is that they likely intended different block styles to also not convey content information.

In other words, if different YAML block styles produce different results in a downstream YAML-processing package (e.g. PyYAML) despite representing the same string content then that seems like a bug in that YAML-processing package.

Anyway, regardless of what the spec says, I tested using PyYAML and it interprets both YAML configurations the same, producing this result in each case:

```python
>>> next(yaml.safe_load_all(open('./test.yaml', 'r')))
{'mapKey': 'fluxcd.io/tag.foo', 'mapValue': '\nregexp:^[0-9]{4}\\.$\n'}
```

**Comment by coderfromhere** [comment]
Oh, you're right, I must've messed with generated yamls and took it as a different PyYAML result. Okay, given PyYAML is a binding to C `libyaml` parser, that behaviour is shared across a few other key frameworks and environments and I can rely on it relatively safely. Thanks for the clarification!

**Comment by Gabriella439** [comment]
You're welcome!
